### DISCUSSION TITLE
Question about updating boundary conditions each time step (depends on time history)

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#22343] [Q&A General] [answered]
Dear Moose community,

I'm new to moose. We would like to implement dynamics rupture earthquake problem. The geometry setup is two rectangular blocks with the bottom boundary of upper block and the top boundary of lower block are considered as fault surface. 

At the current point, we are able to 
1. solve the dynamic FEM problem using "TensorMechanics" "CentralDifference" to get displacement.
2. extract the displacement on the two fault surfaces mentioned above as "AuxVariable" using "AuxKernel" restricted at those surfaces.

Now we would like to do some multiplications with the data on the surfaces(to ensure continuity conditions, which could be done pointwise), and feed the post-computed displacements as boundary condition, then apply on the weak form and solve the system again for the next time step.

I'm wondering which would be the better approach for this. Any help would be appreciated! Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

are you going to be solving this boundary condition implicitly? Or is that part of the system going to be explicit

Guillaume

**Reply by chunhuizhao478** [reply]
Hi Guillaume,

Thanks for your reply. So the system solve is going to be explicit. After getting solution at two surfaces as auxvariable (displacement), for the update of boundary condition, we may experience both situations depend on the model we use:

Slip weakening: the new BC will be calculated explicitly
Rate-and-State Friction: a nonlinear equation using Newton Raphson is needed to be solve.
For both situations, each point has its own set of equation, there is no explicit dependence among different points.

Thanks,
Chunhui

**Reply by GiudGiud** [reply]
Does the BC depend only on local dofs? Or does it have some sort of extended stencil (on the BC sideset, or across the fault?)

**Reply by chunhuizhao478** [reply]
It only depends on local dofs

**Reply by GiudGiud** [reply]
I think you should be able to place all your code in a boundary condition then. 

The two situations could be determined there

**Reply by chunhuizhao478** [reply]
So the current issue we have, due to the limit knowledge of moose, is that we are not sure how to pass a set of aux values into "Function", which then gives an input (as a Function object) for "FunctionDirichletBC" or "FunctionNeumannBC". It would be better if you could direct us to some examples.

**Reply by chunhuizhao478** [reply]
Or you mean "Custom Boundary Conditions" in https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex04_bcs.html ?

**Reply by GiudGiud** [reply]
look at coupled BCs to add in varaible dependence
https://github.com/idaholab/moose/blob/next/framework/src/bcs/CoupledVarNeumannBC.C
Functions do not depend on variables so it's not what you want

**Reply by GiudGiud** [reply]
this example in the tutorial you linked to seems pretty good for what you want to do

**Reply by chunhuizhao478** [reply]
Thanks, I'm looking into it.

**Reply by chunhuizhao478** [reply]
Hi Guillaume, I think the path you directed is promising. Another question is how should we compute elastic forces, basically <stress(displacement), grad(test function)> that is evaluated in the "StressDivergenceTensors" kernel when solving the system, I will need to feed this into Custom Boundary Condition as well. Thanks!

**Reply by GiudGiud** [reply]
The syntax in a BC should be fairly similar to the syntax of a kernel in that regard. 
You ll retrieve the grad_test the same way, and the stress the same way as well as in that kernel

**Reply by chunhuizhao478** [reply]
Hi, Thanks for your reply, we follow this approach and are able to get some results, but with oscillations. We would like to add damping to reduce it (by creating custom kernels, we avoid using Newmark scheme to be able to use central difference), now we want to get stress in the previous time step, we tried to use "getMaterialPropertyOld" in the kernel, but it keep returning the stress in the current step, I'm not sure what we need to do to be able to use this function?

**Reply by GiudGiud** [reply]
are you using automatic differentiation?
are you getting good convergence on each time step (very low residuals) ?
getmaterialPropertyOld should be the way. There is also Older

**Reply by chunhuizhao478** [reply]
Yes, I think we are getting good convergence, that's why we think damping could help. The "getmaterialpropertyold" is not working, do we need custom material kernel to initialize this? I'm sure this function works when I call it in the kernel, but it gives the stress in the current step, which make (stress_current - stress_previous)/dt equals to zero in all the time step. Not sure what we are missing here..

**Reply by mangerij** [reply]
You can use ADTimeDerivativeAux (I forgot the exact name) to compute time derivatives of the stress tensor components and then call them in your Kernels/BCs

**Reply by chunhuizhao478** [reply]
Hi all, thanks for your response. We manage to let the code work using Cohesive Zone Model, but there are something related to the replicated mesh & memory that I would like to ask, maybe create a separate post for it.

**Reply by hugary1995** [reply]
Moose czm doesn't work with distributed mesh.

**Reply by chunhuizhao478** [reply]
Yes, we encounter problems as we increase the domain size running on cluster with EXIT : CODE 9 termination (but smaller domain size works fine, I run the problem on sawtooth2). I'm wondering whether this relates to the replicated mesh since each processor need s to hold all mesh that may cause run out of memory issue? I tried to increase number of cores/ number of nodes but with no luck. (There is a similar post where the issue is fixing by increasing memory), I just want to make sure whether replicated mesh option is a cause or not.

**Reply by hugary1995** [reply]
That indeed sounds like a memory issue caused by replicated mesh based on your description. The czm issue is tracked here https://github.com/idaholab/moose/issues/21161. For now you can stick with replicated mesh, request more nodes, and under subscribe the nodes.

**Reply by hugary1995** [reply]
Aside from that, I am actually surprised that you can get czm to work with an explicit time integrator. Which traction-separation law are you using?

**Reply by chunhuizhao478** [reply]
Hi Gary,

Thanks for your reply. We have submitted a request to the cluster, but could you elaborate more on "request more nodes, and under subscribe the nodes", I'm not quite followed. If this is bounded by the memory of each processor, how the number of nodes would help this case.

For the question, we are using traction-at-split-node approach by Day et al. You could read more here if you are interested: https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2005JB003813

**Reply by hugary1995** [reply]
"request more nodes" would give you more memory, as each node has a fixed amount of memory.
"undersubscribe the nodes" means use fewer cores than the node actually has, which would let you consume less memory by replicating the mesh.

**Reply by hugary1995** [reply]
Oh, node as in "compute node" on a computing cluster, e.g. sawtooth. Not a node in the mesh...

**Reply by chunhuizhao478** [reply]
> "request more nodes" would give you more memory, as each node has a fixed amount of memory.
> "undersubscribe the nodes" means use fewer cores than the node actually has, which would let you consume less memory by replicating the mesh.

I see. Will give a try. Thanks

**Reply by hugary1995** [reply]
Just a disclaimer: I have no idea if your cluster manager will be happy if you undersubscribe the nodes...
